README.md
# jQuery Mory - A Read More Plugin - http://wbotelhos.com/mory

jQuery Mory is a plugin to resume texts along with a read more link.

## License

The jQuery Mory is licensed under [The MIT License](http://www.opensource.org/licenses/mit-license.php)

## Version

	@version        0.1.0
	@since          2012.01.07
	@author         Washington Botelho
	@documentation  wbotelhos.com/mory
	@twitter        twitter.com/wbotelhos

## Required Files

	jquery.mory.min.js

## Default values

	collapseEffect	: undefined
	collapseSpeed	: 150
	collapseText	: 'read less'
	expandEffect	: undefined
	expandSpeed		: 2000
	expandText		: 'read more'
	reticence		: '... '
	size			: 140

## Usage with default values

	$('#text').mory();

	<div id="text"></div>

## Buy me a coffee

You can do it by [PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=X8HEP2878NDEG&item_name=jQuery%20Mory). Thanks! (:


changelog.md
# jQuery Mory - A Read More Plugin - http://wbotelhos.com/mory

### 0.1.0

+ 


spec/spec.js
describe('Using ID', function() {

	beforeEach(function() {
	    $('body').append('<div id="mory">Thomas A. Anderson is a man living two lives. By day he is an average computer programmer and by night a hacker known as Neo. Neo has always questioned his reality, but the truth is far beyond his imagination. Neo finds himself targeted by the police when he is contacted by Morpheus, a legendary computer hacker branded a terrorist by the government. Morpheus awakens Neo to the real world, a ravaged wasteland where most of humanity have been captured by a race of machines that live off of the humans body heat and electrochemical energy and who imprison their minds within an artificial reality known as the Matrix. As a rebel against the machines, Neo must return to the Matrix and confront the agents: super-powerful computer programs devoted to snuffing out Neo and the entire human rebellion.</div>');
	});

	afterEach(function() {
		$('#mory').remove();
	});

	it ('should chainable', function() {
		// given
		var $mory		= $('#mory'),
			className	= 'my-class';

		// when
		$mory.mory().addClass(className);

		// then
	    expect($mory).toHaveClass(className);
	});

	it ('element should create all elements', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory();

		// then
	    expect($mory.children('.mory-resume')).toExist();
	    expect($mory.children('.mory-content')).toExist();
	    expect($mory.children('.mory-content')).toBeHidden();
	    expect($mory.children('.mory-reticence')).toExist();
	    expect($mory.children('.mory-reticence')).toHaveHtml('...&nbsp;');

	    var $more	= $mory.children('.mory-more'),
	    	$link	= $more.children('.mory-link');

	    expect($more).toExist();
	    expect($link).toExist();
	    expect($link).toHaveAttr('href', 'javascript:void(0);');
	    expect($link).toHaveAttr('title', 'read more');
	    expect($link).toHaveHtml('read more');
	});

	it ('number should cut the right length', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory();

		// then
	    expect($mory.children('.mory-resume').html().length).toEqual(140);
	});

	it ('number should cut the right length with custom value', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory({ number: 200 });

		// then
	    expect($mory.children('.mory-resume').html().length).toEqual(200);
	});

	it ('content should expand', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory().children('.mory-more').children('a').click();

		// then
		expect($mory.children('.mory-resume')).not.toBeHidden();
	    expect($mory.children('.mory-content')).not.toBeHidden();
	    expect($mory.children('.mory-reticence')).not.toBeHidden();
	    expect($mory.children('.mory-reticence')).toHaveHtml('&nbsp;');

	    var $more	= $mory.children('.mory-more'),
	    	$link	= $more.children('.mory-link');
	
	    expect($link).toHaveAttr('title', 'read less');
	    expect($link).toHaveHtml('read less');
	});

	it ('content should expand and collapse', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory().children('.mory-more').children('a').click().click();

		// then
		expect($mory.children('.mory-resume')).not.toBeHidden();
	    expect($mory.children('.mory-content')).toBeHidden();
	    expect($mory.children('.mory-reticence')).not.toBeHidden();

	    var $more	= $mory.children('.mory-more'),
	    	$link	= $more.children('.mory-link');
	
	    expect($link).toHaveAttr('title', 'read more');
	    expect($link).toHaveHtml('read more');
	});

	it ('reticence should change it', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory({ reticence: '[..]' });

		// then
	    expect($mory.children('.mory-reticence')).toHaveHtml('[..]');
	});

	it ('reticence should remove it', function() {
		// given
		var $mory = $('#mory');

		// when
		$mory.mory({ reticence: '' });

		// then
	    expect($mory.children('.mory-reticence')).toHaveHtml('&nbsp;');
	});

	it ('should not apply readmore when has no lenght for it', function() {
		// given
		var $mory = $('#mory');

		$mory.html($mory.html().substring(0, 140));

		// when
		$mory.mory();

		// then
		expect($mory.children('.mory-resume')).not.toExist();
	    expect($mory.children('.mory-content')).not.toExist();
	    expect($mory.children('.mory-reticence')).not.toExist();
	});

	it ('number should split the right length', function() {
		// given
		var $mory		= $('#mory'),
			expected	= 'Thomas A. Anderson is a man living two lives. By day he is an average computer programmer and by night a hacker known as Neo. Neo has always';

		// when
		$mory.mory();

		// then
		expect($mory.children('.mory-resume').html()).toEqual(expected);
	});

	it ('should close open tag on resume', function() {
		// given
		var $mory = $('#mory').html('<b>Hello World!</b>');

		// when
		$mory.mory({ number: 5 });

		// then
		expect($mory.children('.mory-resume').html()).toEqual('<b>Hello</b>');
	});

});


js/jquery.mory.js
/*!
 * jQuery Mory - A Read More Plugin - http://wbotelhos.com/mory
 * ------------------------------------------------------------------------
 *
 * jQuery Mory is a plugin to resume texts along with a read more link.
 *
 * Licensed under The MIT License
 *
 * @version        0.1.0
 * @since          2012.01.07
 * @author         Washington Botelho
 * @documentation  http://wbotelhos.com/mory
 * @twitter        http://twitter.com/wbotelhos
 *
 * Usage with default values:
 * ------------------------------------------------------------------------
 * $('#text').mory();
 *
 * <div id="text"></div>
 *
 */

;(function($) {

	var methods = {
		init: function(options) {
			return this.each(function() {

				var self	= this,
					$this	= $(self),
					text	= $.trim($this.html());

				self.opt = $.extend({}, $.fn.mory.defaults, options);

				if ($this.data('mory') || text.length <= self.opt.size) {
					return;
				}

				$this.data('mory', true);

				var size	= methods.getLength.call(this, text),
					resume	= text.slice(0, size);

				text = text.slice(size); 

				self.resume	= $('<span />', { 'class': 'mory-resume', html: resume });
				self.content = $('<span />', { 'class': 'mory-content', html: text }).hide();
				self.reticence = $('<span />', { 'class': 'mory-reticence' });
				self.link = $('<a />', { 'class': 'mory-link', href: 'javascript:void(0);', title: self.opt.expandText, html: self.opt.expandText });
				self.moreLess = $('<span />', { 'class': 'mory-more' }).append(self.link).wrap('<span/>');

				if (self.opt.reticence) {
					self.reticence.html(self.opt.reticence.replace(' ', '&#160;'));
				} else {
					self.reticence.html('&#160;');
				}

				$this.html(self.resume).append(self.content, self.reticence, self.moreLess);

				methods.bindAction.call(this);
			});
		}, bindAction: function() {
			var self = this;

			self.link.toggle(function() {
				$(this).html(self.opt.collapseText).attr('title', self.opt.collapseText);

				if (self.opt.expandEffect == 'fadein') {
					self.content
					.css({ display: 'inline', opacity: 0 })
					.animate({ opacity: 100 }, {
						duration	: self.opt.expandSpeed,
						queue		: false
					});
				} else {
					self.content.css({ 'display': 'inline', 'opacity': 100 });
				}

				self.reticence.html('&#160;');
			}, function() {
				$(this).html(self.opt.expandText).attr('title', self.opt.expandText);

				if (self.opt.collapseEffect == 'fadeout') {
					self.content.animate({ opacity: 0 }, {
						duration	: self.opt.collapseSpeed,
						queue		: false,
						complete	: function() {
							$(this).hide();
							self.resume.css('display', 'inline');
							self.reticence.html(self.opt.reticence);
						}
					});
				} else {
					self.content.hide();
					self.reticence.html(self.opt.reticence);
				}

			});
		}, getLength: function(text) {
			var open	= text.match(this.opt.openTag),
				close	= text.match(this.opt.closeTag),
				single	= text.match(this.opt.singleTag);

			var sum = (open && open.join('').length || 0) + (close && close.join('').length || 0) + (single && single.join('').length || 0);

			return this.opt.size + sum; 
		}, set: function(options) {
			return this.each(function() {
				this.opt = $.extend({}, $.fn.chrony.defaults, $(this).data('options'), options);
			});
		}
	};

	$.fn.mory = function(method) {
		if (methods[method]) {
			return methods[method].apply(this, Array.prototype.slice.call(arguments, 1));
		} else if (typeof method === 'object' || !method) {
			return methods.init.apply(this, arguments);
		} else {
			$.error('Method ' + method + ' does not exist!');
		} 
	};

	$.fn.mory.defaults = {
		collapseEffect	: undefined,
		collapseSpeed	: 150,
		collapseText	: 'read less',
		expandEffect	: undefined,
		expandSpeed		: 2000,
		expandText		: 'read more',
		reticence		: '... ',
		size			: 140,
		openTag			: /(<)([a-zA-Z]+)([^>]*)(>)/g,
		closeTag		: /<(\/)([a-zA-Z])>/g,
		singleTag		: /(<)([a-zA-Z]+)(\s*)(\/>)/g
	};

})(jQuery);
